### DISCUSSION TITLE
Cannot define Many to Many Relationship

### DISCUSSION THREAD

**Discussion by Xinnerpeace** [Budibase/budibase#12544] [Help] [answered]
"Ensure non-key columns are nullable or auto-generated"
But all fields in the through table are either nullable or auto-generated, actually there's only one auto-generated uuid field besides the two foreign key fields.
![image](https://github.com/Budibase/budibase/assets/69332224/1c62e1a6-5646-48f3-b9c1-001d89356e5c)

**Comment by melohagan** [comment]
Hey @Xinnerpeace 

Could you please share the SQL schema for your through table: https://docs.budibase.com/docs/troubleshooting-sql

**Comment by melohagan** [comment, marked as answer]
Hey @Xinnerpeace 

Thanks for the SQL schemas. The issue here is that you have a **Not NULL** field being used as a non-key, non-auto field in your join setup.

So take the setup in your screenshot, and look at these constraints:

<img width="778" alt="Screenshot 2023-12-12 at 08 43 28" src="https://github.com/Budibase/budibase/assets/101575380/79c00311-6543-464e-bd82-bf249432c31d">

If you are using `contacts_id` and `organizations_id` as your keys, then how would Budibase know what to populate **id** with?

On the other hand, if you were to use **id** as one of your relationship keys, then you can see the validation error goes away:

![Screenshot 2023-12-12 at 08 45 00](https://github.com/Budibase/budibase/assets/101575380/4be70a74-dbf0-42ac-8d6d-df6ba33d92bd)

----


I'm guessing what you wanted to have is the following:

```sql
CREATE TABLE public.organizations_contacts (
    contacts_id uuid,
    id serial PRIMARY KEY,
    organizations_id uuid
);

```

<img width="863" alt="Screenshot 2023-12-12 at 08 48 07" src="https://github.com/Budibase/budibase/assets/101575380/24176791-0c4c-420b-9097-96f331b439c0">

![Screenshot 2023-12-12 at 08 52 40](https://github.com/Budibase/budibase/assets/101575380/2d89c64f-eb85-4eb4-b4db-73b7a65b11e5)

**Reply by Xinnerpeace** [reply]
Thx for your reply.
I'm using Directus to manage backend data, the id field is actually auto-generated by Directus and doesn't need Budibase to populate.
On Directus side, I cannot alter the id field, which is a shame.
How do I solve this problem?
If Budibase can provide an option that allow user to still define relationship in this case, but forbid CRUD the relationship, it might be okay.

**Reply by melohagan** [reply]
I don't know much about Directus, but if you find yourself wanting to use Budibase, maybe you can fully migrate across?

**Reply by Xinnerpeace** [reply]
Got a workaround:
Create junction tables in Budibase instead of Directus, and then everything works.
